### ISSUE TITLE
Rendering differences between platforms seems to have increased with 1.3.4

### ISSUE THREAD

**Issue by VirtualParticle** [cashapp/paparazzi#1465]
**Description**
In previous versions, there have been slight variations between the Paparazzi screenshots generated on Linux CI runs and local MacOS runs. This required setting a maximum percent difference above zero, although I was able to set it very low (0.00005%) and still have it work. The primary difference seems to be around text anti-aliasing and color blending more generally.

In the latest version, however, it seems like the difference has increased. It might be something with the new version of layoutlib. Seems like this is happening on the Paparazzi repo itself too: https://github.com/cashapp/paparazzi/pull/1314/files#r1544924645

**Steps to Reproduce**

Here's an example of a test with a screenshot originally generated on MacOS and failing on Linux with Paparazzi 1.3.4 (this has a difference of 0.007134%):

```kotlin
@Test
fun example() {
    paparazzi.snapshot {
        Box(
            modifier = Modifier.background(Color(0xFF000033))
        ) {
            Text("ExampleText", color = Color.White)
        }
    }
}
```

![image](https://github.com/cashapp/paparazzi/assets/20135416/21c5b45c-9f64-4c7d-89d1-052e0271145d)

With Paparazzi 1.3.3, the same test passes on Linux CI with the screenshot generated on MacOS, even with the `maxPercentDifference` set to 0.

When the accessibility overlay is enabled, the error increases to 0.083399%.

![image](https://github.com/cashapp/paparazzi/assets/20135416/e89b046d-54c3-460c-8181-d50abe7311ed)

**Expected behavior**
I'd expect the tests to be identical on all platforms but that wasn't the old behavior either. Acceptable behavior would just be to have the discrepancy be small enough to ignore. To deal with the higher difference scenarios, we'd have to set the `maxPercentDifference` to be high enough where actual bugs may be missed.

**Additional information:**
- Paparazzi Version: 1.3.4
- OS: MacOS (local) and Linux (CI)
- Compile SDK: 34
- Gradle Version: 8.7
- Android Gradle Plugin Version: 8.4.1

**Comment by addonsdetector** [comment]
We are having the same issue. PR to bump to 1.3.4 fails on CI, while on our developer Macbooks  no `verify` step fails and also `record` step does not generate updated screenshot. So this blocks us from going to 1.3.4, cause we need this in our CI.

**Comment by peterdk** [comment]
We are having the same issue. PR to bump to 1.3.4 fails on CI, while on our developer Macbooks  no `verify` step fails and also `record` step does not generate updated screenshot. So this blocks us from going to 1.3.4, cause we need this in our CI.

**Comment by jrodbx** [comment]
> So this blocks us from going to 1.3.4, cause we need this in our CI.

Have you experimented with values for [maxPercentDifference](https://github.com/cashapp/paparazzi/blob/master/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt#L36)?

**Comment by hanna-h** [comment]
> > So this blocks us from going to 1.3.4, cause we need this in our CI.
> 
> Have you experimented with values for [maxPercentDifference](https://github.com/cashapp/paparazzi/blob/master/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt#L36)?

As @VirtualParticle rightfully mentioned:

> To deal with the higher difference scenarios, we'd have to set the maxPercentDifference to be high enough where actual bugs may be missed.

**Comment by jrodbx** [comment]
understood.  do @peterdk and @VirtualParticle work on the same project?  if not, i'd be curious to know if the maxPercentDifference might work for @peterdk.

**Comment by jrodbx** [comment]
> As @VirtualParticle rightfully mentioned:
> 
> > To deal with the higher difference scenarios, we'd have to set the maxPercentDifference to be high enough where actual bugs may be missed.

understood, but do @peterdk and @VirtualParticle work on the same project? if not, i'd be curious to know if the maxPercentDifference might work for @peterdk.

**Comment by VirtualParticle** [comment]
> > As @VirtualParticle rightfully mentioned:
> > > To deal with the higher difference scenarios, we'd have to set the maxPercentDifference to be high enough where actual bugs may be missed.
> 
> understood, but do @peterdk and @VirtualParticle work on the same project? if not, i'd be curious to know if the maxPercentDifference might work for @peterdk.

We do not work on the same project.

**Comment by mshearer123** [comment]
Same issue, different project.  Seems mostly on bottom sheets and progress indicator. 
have tried the animation rule

**Comment by peterdk** [comment]
> Have you experimented with values for [maxPercentDifference](https://github.com/cashapp/paparazzi/blob/master/paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt#L36)?
I'm very reluctant to start working around image differences by raising a error threshold. Especially if it's caused by OS differences.

**Comment by BraisGabin** [comment]
With Paparazzi 1.3.2 I had a suite of 1428 tests passing with a `maxPercentDifference = 0`. When bumping to 1.3.4 1344 of those 1428 fail because of the difference between the screenshots generated on Mac and the ones generated on Linux. Even moving the `maxPercentDifference` to `0.01` I had some of those tests failing.

**Comment by peterdk** [comment]
@BraisGabin did you do a re-record after bumping? Cause there are (unfortunately) sometimes minor changes when bumping paparazzi. Cause 1428 failures is quite high, to only link that to the difference between mac and linux.

**Comment by BraisGabin** [comment]
Yes, I did.

As I said on 1.3.2 I was running all my tests with `maxPercentDifference = 0` without any problem. With 1.3.2 I always got green tests recording on Mac and then verifying on Linux.

Now with 1.3.4 I got all those failing tests. If I change the `maxPercentDifference` to `0.01` I get less failing tests but I still get some.

**Comment by dniHze** [comment]
It's the same story for us at Cuvva with loads of failures between Mac/Arm64 and Linux/x86. Just like others, we got loads of screenshots regenerated after the upgrade (from 1.3.1 to 1.3.4). I'm not sure if it has something to do with Paparazzi itself, as it could be the layoutlib doing something wonky on different platforms/architectures.

**Comment by linean** [comment]
Same issue here; the max difference is about 0.2%. Unfortunately, that's too high and could introduce bugs. Our ideal range was 0.05%.

**Comment by ReginFell** [comment]
We have pretty much the same issue. After upgrading from 1.3.3 to 1.3.4, a lot of screenshots were updated on Mac, but they fail on the CI (Linux). With `maxPercentDifference` = 0.005, they fail, but with `maxPercentDifference` = 0.1, it works, although it creates a huge threshold for issues.

**Comment by savvisingh** [comment]
Having the same issue, is there any other solution rather than increasing maxPercentDifference, Can we expect a fix for this?

**Comment by BraisGabin** [comment]
> Having the same issue, is there any other solution rather than increasing maxPercentDifference, Can we expect a fix for this?

My work around, right now, is just not upgrade to this version.

**Comment by jrodbx** [comment]
> Can we expect a fix for this?

I'm not sure what the root cause is yet, so hard to promise a fix yet, but I'll be investigating this issue this week.

**Comment by gabrielittner** [comment]
@jrodbx I've got a very simple reproducer for a platform difference, just having a semi transparent box.
```
Column(
  Modifier
    .background(Color.White)
    .fillMaxSize()
) {
  Box(
    Modifier
      .fillMaxSize()
      .padding(48.dp)
      .background(Color.Black.copy(alpha = 0.5f))
  ) {
  }
}
```

I've put it into a test and the CI fails on Linux but succeeds on macOS (the committed screenshot was generated on macOS):
https://github.com/cashapp/paparazzi/commit/6f50b720c2e086f3457905c68730d37ad085d4b5
https://github.com/freeletics/paparazzi/actions/runs/9592840917

**Comment by jrodbx** [comment]
FYI, still investigating, but wanted to share notes:

For this investigation, I've been using the snippet provided in https://github.com/cashapp/paparazzi/issues/1465#issue-2326059848 and had already been investigating before @gabrielittner's [simpler repro here](https://github.com/cashapp/paparazzi/issues/1465#issuecomment-2179960990).

First, I ran [the tests in this branch](https://github.com/cashapp/paparazzi/commit/3a400ed6152ecaba0c82bf2d4d230a0d50c57595#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721) on JDK {17,18} x {linux, mac, win} x {legacy view, compose} toolkits, using the version @ SHA = 7ca1390.  This is the SHA right before Iguana merged.

All 12 files (2 jdks x 3 os x 2 toolkits) are identical.  This is where we want to be.

After Iguana, JDKs still have no impact (whew...), but OS platforms clearly do.

### Theory 1

Hypothesis: [This commit](https://android.googlesource.com/platform/frameworks/layoutlib/+/8fc03892900c62c81f47310d2fe706b0c024143d%5E%21/#F0) (Use premultiplied alpha images for layoutlib) is the root cause for the rendering change
Result: FALSE
While analyzing the decompiled class for RenderSessionImpl in Iguana's LayoutLib shows that the BufferedImage type has changed from type 2 (TYPE_INT_ARGB) to type 3 (TYPE_INT_ARGB_PRE), supplying an ImageTransformation via SessionParams in order to reverse this by copying the type 2 image into a new image of type 3 and diffing the two shows that the two images are *still identical* (presumably because the image has no actual alpha despite support for it?)

### Theory 2

Hypothesis: An additional change is needed to account for recomposition which would allow pending work to stabilize.
Result: FALSE
This was a bit of a guess attempt, given that I know @geoff-powell had been looking into this.  I cherry-picked from [Geoff’s branch](https://github.com/cashapp/paparazzi/commit/f2d2af52bcf47eda16ec02edb50ceadc64e7bc27) which attempts to solve this.  However, it does not impact the affected images for this repro.  In retrospect, this was a silly trial and the outcome makes sense as the repro snippet does not appear to do anything special that would benefit from a second layout pass.

### Theory 3

Hypothesis: The view hierarchies' metadata are slightly different, so different inputs result in different outputs.
Result: FALSE
TL;DR I copied the sources of https://github.com/square/radiography and https://github.com/square/curtains into a branch and analyzed the hierarchies.  This was the point where I also decided to write a legacy view (vs a Composable version of the test) since legacy view hierarchies seemed easier to analyze recursively to peek at its metadata.

### Theory 4 (ongoing)

Hypothesis: The way in which the hierarchy gets rasterized is the problem.

[This commit](https://android.googlesource.com/platform/frameworks/layoutlib/+/ebdd83e4be7e8d89a38e3f316b2e15112f61ca30%5E%21/#F5) fundamentally changed out drawing occurs in Layoutlib from Hedgehog to Iguana.  

When Paparazzi is ready to render the image from a view hierarchy, it calls:
```
// PaparazziSdk#L302 
val result = renderSession.render(true)
```
which calls:
```
// RenderSessionImpl#252
return this.renderAndBuildResult(freshRender, false);
```

Later down the road, when the native image has been captured, it's copied into the image buffer that we ultimately write to the file system:

```
// RenderSessionImpl#renderAndBuildResult
        int[] imageData = ((DataBufferInt)this.mImage.getRaster().getDataBuffer()).getData();
        Image.Plane[] planes = this.mNativeImage.getPlanes();
        IntBuffer buff = planes[0].getBuffer().asIntBuffer();
        int len = buff.remaining();
        buff.get(imageData, 0, len);
```

In a debugger, viewing `imageData` in the local stack frame shows the captured bitmap screenshot.

What happens in between these two points differs from Hedgehog to Iguana...

### Pre-Iguana
```
// RenderSessionImpl#347
        renderResult = renderAndBuildResult(this.mViewRoot, this.mRenderer);
```
```
    private Result renderAndBuildResult(ViewGroup viewRoot, HardwareRenderer renderer) {
        AttachInfo_Accessor.dispatchOnPreDraw(viewRoot);
        RenderNode node = viewRoot.updateDisplayListIfDirty();
        renderer.setContentRoot(node);
        renderer.createRenderRequest().syncAndDraw();
        return Status.SUCCESS.createResult();
    }
```
```
// LayoutlibRenderer#21
    public void setContentRoot(RenderNode content) {
        RecordingCanvas canvas = this.mRootNode.beginRecording();
        canvas.scale(this.scaleX, this.scaleY);
        canvas.enableZ();
        canvas.drawColor(0, BlendMode.CLEAR);
        if (content != null) {
            canvas.drawRenderNode(content);
        }

        canvas.disableZ();
        this.mRootNode.endRecording();
    }
```

### Post-Iguana

```
// RenderSessionImpl#335
        this.mRenderer.draw(this.mViewRoot);
```
```
// LayoutlibRenderer#25
        this.draw(viewGroup, rootView.mAttachInfo, new ThreadedRenderer.DrawCallbacks() {
```
```
// ThreadedRenderer#374
    void draw(View view, View.AttachInfo attachInfo, DrawCallbacks callbacks) {
        attachInfo.mViewRootImpl.mViewFrameInfo.markDrawStart();
        this.updateRootDisplayList(view, callbacks);
```
```
// ThreadedRenderer#355
    private void updateRootDisplayList(View view, DrawCallbacks callbacks) {
        ...
        this.updateViewTreeDisplayList(view);
        ...
            RecordingCanvas canvas = this.mRootNode.beginRecording(this.mSurfaceWidth, this.mSurfaceHeight);

            try {
                int saveCount = canvas.save();
                canvas.translate((float)this.mInsetLeft, (float)this.mInsetTop);
                callbacks.onPreDraw(canvas);
                canvas.enableZ();
                canvas.drawRenderNode(view.updateDisplayListIfDirty());
                canvas.disableZ();
                callbacks.onPostDraw(canvas);
                canvas.restoreToCount(saveCount);
                this.mRootNodeNeedsUpdate = false;
            } finally {
                this.mRootNode.endRecording();
            }
        ...
```

TL;DR as the commit message calls out, more of the execution flow is moved out of the Layoutlib shim code and into the actual Android rendering code, which is theoretically the ideal approach.  But somehow, that also appears to be causing the rendering differences...?

I would ideally like to debug the canvas draw native calls that call into the layoutlib native dylib on Mac OS, but I don't know and I think it involves having the debug symbols?  And I haven't successfully been able to build Layoutlib from source to maybe work around that.  Any advice from the community on how to achieve this?

In the meantime, using the Paparazzi test rule as an Android platform bootstrap (thanks to Layoutlib), I've found that I can write to a RecordingCanvas and create RenderNodes directly in a unit test, so I may try inlining the above code to see if I can come up with a minimal repro to file against Google's Layoutlib team in request for more info.

**Comment by jrodbx** [comment]
@gabrielittner one theory I also wanted to try/dismiss was whether I messed up the Iguana migration. I recall you had a commit where you had tweaked the `withTime()` argument with a bespoke offsetting of Choreographer_Delegate.sChoreographerTime in order to get your version working.  I ultimately went with a different approach, but if you still have that branch/commit somewhere (Github history has since cleaned it up from your original PR), it would be worth trying out.

**Comment by kozaxinan** [comment]
We had the exact background color with alpha usage. We mitigated the paparazzi issue using brush and alpha instead of color with alpha. It is not great to change production code to make the test happy but it is better than ignoring the tests

I wanted to share this working alpha version. It might help you create different hypotheses. 

```
  .background(
      brush = SolidColor(Color.Black),
      alpha = 0.5f,
  )
```

**Comment by gabrielittner** [comment]
@jrodbx this was the last state of my branch from before your iguana PR was merged https://github.com/cashapp/paparazzi/compare/master...freeletics:paparazzi:layoutlib-iguana It did also have platform differences though as you can see in the last commit

**Comment by jrodbx** [comment]
> this was the last state of my branch from before your iguana PR was merged

Ok, confirmed that this branch would have resulted in the same image binary data as what has been merged, so we can rule that out as a possible reason.

**Comment by jrodbx** [comment]
> @jrodbx I've got a very simple reproducer for a platform difference, just having a semi transparent box.
> ...
> I've put it into a test and the CI fails on Linux but succeeds on macOS (the committed screenshot was generated on macOS): [6f50b72](https://github.com/cashapp/paparazzi/commit/6f50b720c2e086f3457905c68730d37ad085d4b5) https://github.com/freeletics/paparazzi/actions/runs/9592840917

@gabrielittner i've incorporated your repro and that of @kozaxinan into my two working branches (please check my work!):
* [jrod/2024-06-18/pre-iguana](https://github.com/cashapp/paparazzi/commit/2260dee4bd983d63c63ef850943f8a364e8dd697#diff-fc48259930bcc0f562ed494438179ef8d0f1cfe1d2adeda963ae208b6b5a8d05)
* [jrod/2024-06-18/post-iguana](https://github.com/cashapp/paparazzi/commit/3e51b8fb4c41fd9c4de785b549bc9cfd9222c153#diff-fc48259930bcc0f562ed494438179ef8d0f1cfe1d2adeda963ae208b6b5a8d05)

...and I'm not seeing any differences in the outputs across OS platforms when I test with pre-Iguana and post-Iguana bases.

One thing to call out is in both of my test branches, I commented out scaling of the bitmap to focus on the Iguana effects themselves.  Could that be what's accounting for the differences in your respective observations?

**Comment by jrodbx** [comment]
Ok, it's been a while. Here's an update on what I've found:

Over the past two weeks, I've been running several experiments to understand the pre- and post-Iguana changes.  A week into it, I backtracked and decided to disable both scaling and max % difference and to re-analyze all screenshots to better understand the accuracy of verification pre-Iguana.

After updating all golden images to full resolution and applying pixel-perfect comparisons, 7 tests still [failed on Ubuntu and Windows CI shards](https://github.com/cashapp/paparazzi/actions/runs/9738818469/job/26872963439): 

- AccessibilityRenderExtensionTest.verify baseline
  - java.lang.AssertionError: Images differ (by 0.000001%)
- AccessibilityRenderExtensionTest.test without layout params set
  - java.lang.AssertionError: Images differ (by 0.000001%)
- AccessibilityRenderExtensionTest.verify changing view hierarchy order doesn't change accessibility colors
  - java.lang.AssertionError: Images differ (by 0.000001%)
- PaparazziPluginTest.verifyFailureMultiModule
  - java.lang.AssertionError: Expected % diff less than 0.0, but was: 0.0016120491406439118
- PaparazziPluginTest.configIsUpdatable
  - java.lang.AssertionError: Expected % diff less than 0.0, but was: 2.465549548580292E-6
- PaparazziPluginTest.verifyFailure
  - java.lang.AssertionError: Expected % diff less than 0.0, but was: 0.0016120491406439118
- PaparazziPluginTest.widgets
  - java.lang.AssertionError: Expected % diff less than 0.0, but was: 3.573376407660471E-5

4 of these, namely:
- AccessibilityRenderExtensionTest.verify baseline
- AccessibilityRenderExtensionTest.test without layout params set
- AccessibilityRenderExtensionTest.verify changing view hierarchy order doesn't change accessibility colors
- PaparazziPluginTest.widgets

...fail for similar reasons, i.e., cross-platform gradient rendering differences (dithering techniques?)

<img width="691" alt="Screenshot 2024-06-29 at 6 57 41 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/55e57f51-0bf1-4d92-ac5e-f0203abfc08f">
<img width="1560" alt="Screenshot 2024-06-30 at 6 33 29 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/725ccb42-88df-417b-a77d-23eba516d6c9">

I was tempted to see how different these pixels were, so I compared their RGB values.

<img width="1392" alt="Screenshot 2024-06-29 at 7 16 27 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/4ba20e01-b912-48a7-929a-a952396ea307">

Seeing the consistent off-by-one errors led me to file https://issuetracker.google.com/issues/350075298, under the impression that this is likely some rounding error in the Skia rendering engine.

Moving onward, a similar phenomenon occurred in:
PaparazziPluginTest.configIsUpdatable

<img width="1558" alt="Screenshot 2024-06-30 at 6 33 51 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/086a9ca5-f78c-4af8-b9d6-19dd47cc1bdf">


Again, all different pixels are off by one.

So, I *reworked the diffing logic in ImageUtils to allow off-by-one RGB component values when comparing images*, allowing those 4 tests to pass.

That leaves us with:
PaparazziPluginTest.verifyFailure
PaparazziPluginTest.verifyFailureMultiModule 

The root cause here?  The "Expected" and "Actual" labels we apply to the delta diffs :)

After spending a day and half failing miserably trying to get Graphics2D APIs to render those simple text labels in a pixel-perfect fashion, it turns out:
1) Disable antialiasing on MacOSX 14+ just isn't possible
2) Graphics2D defers to the native text rendering engines so perfect alignment is near impossible

I considered disabling the printing of the text labels for tests, similar to what @drury did in [insert APNG PR here], but leaned towards another idea, especially after another member of the community suggested the same idea -- pre-rendering the text labels into placeholder image labels:

<img width="561" alt="Screenshot 2024-07-06 at 8 53 42 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/abe1d56b-633d-46bb-8b9b-1f53fb46a096">


Cool, all tests pass! Here's a [cleaned-up branch highlighting these pre-Iguana learnings](https://github.com/cashapp/paparazzi/commits/jrod/2024-06-30/pre-iguana/) so far.

Now, how does Iguana change things?

Well, one obvious thing is Iguana Layoutlib bumps the compile sdk to API 34, which introduced [non-linear font scaling](https://developer.android.com/about/versions/14/features#non-linear-font-scaling).

Less obvious subtleties were changes to text antialiasing and vector graphics renderings:

<img width="1590" alt="Screenshot 2024-07-06 at 8 43 34 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/370ad077-6e5d-4b5e-9488-2d58b9050694">

<img width="1590" alt="Screenshot 2024-07-06 at 4 33 10 PM" src="https://github.com/cashapp/paparazzi/assets/1868149/5cdad7bc-a5ca-47d2-96dc-1b8aa29d0db1">

Again, comparing pixel rgb values showed that resolving this by [relaxing the differ even further](https://github.com/cashapp/paparazzi/commit/7c13558951d079b2ccdbaf4843a187b36704c50f) to allow rgb component values to be “off-by-two”.

Furthermore, this allows us to [remove this workaround](https://github.com/cashapp/paparazzi/blob/9fa26b38d5929089a8c21e603ecc56b3bda79ef9/paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt#L710-L712) added in the original Iguana bump.

While this covers up to the Iguana bump, I figured I’d keep on examining if there were any other issues introduced in 1.3.4.  Only two commits slightly changed things.

1) Support for [APNG-based verifications](https://github.com/cashapp/paparazzi/commit/1f7ae6b5336075822552bbab4b00e5b3598d78c) (i.e., gif() support) showed no pixel data differences, but unscaled image file sizes went up substantially.
  - [paparazzi-gradle-plugin/src/test/projects/appcompat-missing/src/test/resources/arrow_missing.png](https://github.com/cashapp/paparazzi/commit/1f7ae6b5336075822552bbab4b00e5b3598d78c8#diff-06e44fdb5f6863dd09b7ba1b7f9f6e78f03176c80e76c9698f7aa6fe55267008) by 12.78% 
<img width="1593" alt="Screenshot 2024-07-07 at 12 45 13 AM" src="https://github.com/cashapp/paparazzi/assets/1868149/860f4c89-79d6-4c58-b087-86766c6c695a">

  - [paparazzi-gradle-plugin/src/test/projects/custom-fonts-code/src/test/resources/textviews.png](https://github.com/cashapp/paparazzi/commit/cb40d7cc03a7f9f2069cac91c73608b8d7adbf28#diff-af6c48d80cb8dda6aeb6cd8492cf7c4da542c42e0072130f664aa1a9593febd0) by 20.15%
<img width="1595" alt="Screenshot 2024-07-07 at 12 48 12 AM" src="https://github.com/cashapp/paparazzi/assets/1868149/e0b2592a-9dbc-494e-bc39-050ae4a76e95">

  - Not a showstopper, but would be a nice exercise to see how we can return to the original file sizes or even improve size further.

2) Bumping [androidx.compose.material:material from v1.5.4 to v1.6.6](https://github.com/cashapp/paparazzi/commit/0996086ec2b791abffc54c30d094c16ae6e5fed3) resulted in [acceptable changes](https://github.com/cashapp/paparazzi/commit/dc0ce7bcc2f0daea378ca0cb758e66daafc15ba5) due to the change in the default value of include font padding: https://medium.com/androiddevelopers/fixing-font-padding-in-compose-text-768cd232425b

After all this, I wanted to see if reapplying image scaling to 1000px would break any of the applied fixes.  [It did not.](https://github.com/cashapp/paparazzi/commit/e9a109c84bb11d456f6e90e7670fe74dc0ac89d1)

So...here's a [cleaned-up branch highlighting everything in 1.3.4](https://github.com/cashapp/paparazzi/commits/jrod/2024-06-30/pre-iguana/).

**Comment by jrodbx** [comment]
In summary, here are some action items:

Immediate:
* Extract the diffing logic in ImageUtils into some sort of differ and relax its strictness/similarity checking, i.e., it will allow rgb components that differ by <= 2 to be considered similar.  In the current set of tests, this appears to overcome cross-platform rendering differences due to antialiasing (for text) and dithering (for gradients).  With this, we can remove the max percent difference parameter!  It was never really a great option, but served as a quick bandaid for the types of problems that diffing should now solve.

* Rethink how we render the delta diff.  Before realizing that pixel-perfect text renderings across platforms is a pipe dream, I attempted to do so, by tweaking Graphics2D.setRenderingHints and even trying to coerce execution paths on MacOSX14+ to act like it wasn't MacOSX14+.  For now, we'll just pre-render the label strings to image files and compose that into the delta diff, but a better approach might be to separate the before/after/delta images and instead display them in an html table and avoid hacking around with Graphics2D altogether.

Afterwards:
* Experiment with other differ algorithms.  I am doubtful that the "off by two" approach will solve all problems; the androidx team had experimented with an implementation that uses the SSIM algorithm.  @romainguy has also suggested experimenting with implementations in FLIP and SIFT.

* Refactor most plugin tests to use verifyPaparazziDebug vs testDebug.  This should obviate ImageSubject and TestSnapshotVerifier, each of which required me to remember to tweak them when experimenting with the differ solution.

* See if Layoutlib can improve in its renderings.  While Graphics2D delegates to the native UI toolkit, Layoutlib switched over to Android's Skia rendering pipeline back in Android Studio 4.1 (pre-Arctic Fox) which runs on FreeText/OpenGL/Vulkan.  So why are gradients and text not able to rendering pixel perfect?  Is the off-by-[one, two] issue a bug with a simple fix (say, like platform-specific rounding errors)?

**Comment by eboudrant** [comment]
🚢 it ... when can we expect paparazzi 1.3.5? :)

**Comment by SimonMarquis** [comment]
👋 Hi, out of curiosity, is there a way to remove these `actual` & `expected` labels?
For us, they provide no useful information since we are used to have before/after images next to each other, and if the changed pixels were located below those label, they would hide it.
The `ApngVerifier.withErrorText` seems be what I'm looking for, but it is not exposed as a public API (and always true in the default configuration).

**Comment by TWiStErRob** [comment]
Probably worth a separate issue

**Comment by VirtualParticle** [comment]
Unfortunately I'm still seeing this issue on 1.3.5, although not to the same extent. I'll see if I can get a reproducible sample and see exactly how far off the pixels are.
